File: AtCoder/KyoproTenke/python/012-Red_Painting.py
H, W = map(int, input().split())

board = [[False] * W for _ in range(H)]
par = list(range(H * W))
rnk = [0] * (H * W)

di = [0, 1, 0, -1]
dj = [1, 0, -1, 0]


def find(u):
    global par
    if par[u] == u:
        return u
    par[u] = find(par[u])
    return par[u]


def unite(u, v):
    u = find(u)
    v = find(v)
    if u == v:
        return
    if rnk[u] < rnk[v]:
        par[u] = v
    else:
        par[v] = u
        if rnk[u] == rnk[v]:
            rnk[u] += 1


def same(u, v):
    return find(u) == find(v)


def vertex(i, j):
    return i * W + j


Q = int(input())
for _ in range(Q):
    t, *q = map(int, input().split())
    if t == 1:
        r, c = map(lambda x: x - 1, q)
        board[r][c] = True
        for _di, _dj in zip(di, dj):
            _r = r + _di
            _c = c + _dj
            if 0 <= _r < H and 0 <= _c < W and board[_r][_c]:
                unite(vertex(_r, _c), vertex(r, c))
    else:
        ra, ca, rb, cb = map(lambda x: x - 1, q)
        if board[ra][ca] and board[rb][cb] and same(vertex(ra, ca), vertex(rb, cb)):
            print("Yes")
        else:
            print("No")

File: AtCoder/KyoproTenkei/cpp/012-Red_Painting.cpp
#include <bits/stdc++.h>
using namespace std;
using ll = long long;

const ll MAX_H = 2000;
const ll MAX_W = 2000;
const ll di[] = {0, 1, 0, -1};
const ll dj[] = {1, 0, -1, 0};
ll H, W, Q;
bool board[MAX_H][MAX_W];

ll par[MAX_H * MAX_W], rnk[MAX_H * MAX_W];

void init() {
  for (ll i = 0; i < H * W; ++i) {
    par[i] = i;
    rnk[i] = 0;
  }
}

ll find(ll u) {
  if (par[u] == u) return u;
  return par[u] = find(par[u]);
}

void unite(ll u, ll v) {
  u = find(u);
  v = find(v);
  if (u == v) return;
  if (rnk[u] < rnk[v])
    par[u] = v;
  else {
    par[v] = u;
    if (rnk[u] == rnk[v]) ++rnk[u];
  }
}

bool same(ll u, ll v) {
  return find(u) == find(v);
}

ll vertex(ll i, ll j) {
  return i * W + j;
}

int main() {
  cin >> H >> W >> Q;
  init();
  while (Q--) {
    ll t;
    cin >> t;
    if (t == 1) {
      ll r, c;
      cin >> r >> c;
      --r, --c;
      board[r][c] = true;
      for (ll k = 0; k < 4; ++k) {
        ll _r = r + di[k];
        ll _c = c + dj[k];
        if (0 <= _r && _r < H && 0 <= _c && _c < W && board[_r][_c])
          unite(vertex(r, c), vertex(_r, _c));
      }
    } else {
      ll ra, ca, rb, cb;
      cin >> ra >> ca >> rb >> cb;
      --ra, --ca, --rb, --cb;
      if (board[ra][ca] && board[rb][cb]
          && same(vertex(ra, ca), vertex(rb, cb)))
        puts("Yes");
      else
        puts("No");
    }
  }
}